src/DataTypes/Serializations/SerializationObject.cpp

```diff
@@ -474,7 +474,10 @@ void SerializationObject::deserializeBinaryBulkStatePrefix(
         /// Wait for all tasks to be executed.
         std::exception_ptr exception;
         for (const auto & task : tasks)
-            exception = task->wait();
+        {
+            if (auto e = task->wait())
```

Comment:
Why not to use ThreadPoolCallbackRunnerLocal for scheduling tasks to the thread pool? It can do all necessary tricks with attaching/detaching to/from a thread group + exception handling is much more convenient there.

Reply:
Because it doesn't allow to steal tasks from the thread pool. Here if the thread pool is busy with lots of tasks current thread steals and executes some tasks.

Reply:
Is it a shared thread pool? Because if not I have a strong feeling that this scheme doesn't worth it. We had the similar problem with backups and saw the situation when all the threads from the threadpool were waiting for the main thread to finish the execution of "stolen" task and continue scheduling other ones back to the threadpool.

Reply:
> Is it a shared thread pool?

It's a global thread pool

Reply:
Without tasks stealing we could also get a deadlock here when we have nested JSONs, because in this case task from threadpool can schedule new tasks in the same pool